src/query.c

```diff
@@ -1753,63 +1776,91 @@ int QueryNode_ForEach(QueryNode *q, QueryNode_ForEachCallback callback, void *ct
+                                            .valLen = attr->vallen };
+    vq->params.params = array_ensure_append_1(vq->params.params, param);
+    bool resolve_required = false;  // at this point, we have the actual value in hand, not the query param.
+    vq->params.needResolve = array_ensure_append_1(vq->params.needResolve, resolve_required);
+    return 1;
+  }
+  return 0;
+}
+
 static int QueryNode_ApplyAttribute(QueryNode *qn, QueryAttribute *attr, QueryError *status) {
 
 #define MK_INVALID_VALUE()                                                         \
   QueryError_SetErrorFmt(status, QUERY_ESYNTAX, "Invalid value (%.*s) for `%.*s`", \
                          (int)attr->vallen, attr->value, (int)attr->namelen, attr->name)
 
+  int res = 0;
   // Apply slop: [-1 ... INF]
-  if (STR_EQCASE(attr->name, attr->namelen, "slop")) {
+  if (STR_EQCASE(attr->name, attr->namelen, SLOP_ATTR)) {
     long long n;
     if (!ParseInteger(attr->value, &n) || n < -1) {
       MK_INVALID_VALUE();
-      return 0;
+      return res;
     }
     qn->opts.maxSlop = n;
+    res = 1;
 
-  } else if (STR_EQCASE(attr->name, attr->namelen, "inorder")) {
+  } else if (STR_EQCASE(attr->name, attr->namelen, INORDER_ATTR)) {
     // Apply inorder: true|false
     int b;
     if (!ParseBoolean(attr->value, &b)) {
       MK_INVALID_VALUE();
-      return 0;
+      return res;
     }
     qn->opts.inOrder = b;
     qn->opts.flags |= QueryNode_OverriddenInOrder;
+    res = 1;
 
-  } else if (STR_EQCASE(attr->name, attr->namelen, "weight")) {
+  } else if (STR_EQCASE(attr->name, attr->namelen, WEIGHT_ATTR)) {
     // Apply weight: [0  ... INF]
     double d;
     if (!ParseDouble(attr->value, &d) || d < 0) {
       MK_INVALID_VALUE();
-      return 0;
+      return res;
     }
     qn->opts.weight = d;
+    res = 1;
 
-  } else if (STR_EQCASE(attr->name, attr->namelen, "phonetic")) {
+  } else if (STR_EQCASE(attr->name, attr->namelen, PHONETIC_ATTR)) {
     // Apply phonetic: true|false
     int b;
     if (!ParseBoolean(attr->value, &b)) {
       MK_INVALID_VALUE();
-      return 0;
+      return res;
     }
     if (b) {
       qn->opts.phonetic = PHONETIC_ENABLED;  // means we specifically asked for phonetic matching
     } else {
       qn->opts.phonetic =
           PHONETIC_DISABLED;  // means we specifically asked no for phonetic matching
     }
+    res = 1;
     // qn->opts.noPhonetic = PHONETIC_DEFAULT -> means no special asks regarding phonetics
     //                                          will be enable if field was declared phonetic
 
-  } else {
-    QueryError_SetErrorFmt(status, QUERY_ENOOPTION, "Invalid attribute %.*s", (int)attr->namelen,
-                           attr->name);
-    return 0;
+  } else if (STR_EQCASE(attr->name, attr->namelen, YIELD_DISTANCE_ATTR) && qn->opts.flags & QueryNode_YieldsDistance) {
+    qn->opts.distField = rm_strndup(attr->value, attr->vallen);
+    res = 1;
+
+  } else if (qn->type == QN_VECTOR) {
```

Comment:
Why do we need to limit this for vector only. 
We can set the flag regardless and just ignore it if not supported.

Reply:
Note that in `QueryVectorNode_ApplyAttribute` we set the vector attributes into `VectorQueryParams` struct, which is part of a vector node only. To follow your suggestion, I guess we should add the vector attributes to the generic `QueryNodeOptions`. Let's discuss it...